cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go

```diff
@@ -708,3 +725,47 @@ func (c *machineController) allowedByAutoDiscoverySpecs(r *unstructured.Unstruct
 
 	return false
 }
+
+// Get an infrastructure machine template given its GVR, name, and namespace.
+// This function exists so that we can implement a cache for the
+// infrastructure template objects. A local cache is used in favor of the
+// client-go informers because of the frequency of read requests combined
+// with the fact that the infrastructure template will not be updated
+// frequently (if at all). In practice, the autoscaler can make calls to
+// the API server at a high frequency, which results in many calls being
+// throttled on the client-side. To reduce the amount of throttling we keep
+// an in-memory cache of infrastructure templates. The cache has a time to
+// live on any entry of 1 minute.
+func (c *machineController) getInfrastructureResource(resource schema.GroupVersionResource, name string, namespace string) (*unstructured.Unstructured, error) {
+	// this cache key is only used internally, join the GVR+name+namespace
+	cachekey := strings.Join([]string{resource.String(), name, namespace}, "")
+	if entry, ok := infrastructureTemplateCache[cachekey]; ok {
+		// if the infrastructure template exists in the cache and is under the time to live, return it
+		if time.Now().Before(entry.timeToLive) {
+			return entry.template, nil
+		}
+	}
+
```

Comment:
>because of the frequency of read requests

It's not clear to me why we need this handcrafted cache, could you elaborate on the comment above?

Shouldn't the managementClient just be told what it caches and what it doesn't via e.g controller runtime primitives? e.g
```
NewCache: cache.BuilderWithOptions(cache.Options{
			SelectorsByObject: cache.SelectorsByObject{
...
```

Reply:
good question. i initially tried adding the infrastructure template types to our dynamic client, but what i noticed when running that way is that we get client side throttled several times per sync loop. given that these records will not update very frequently, if at all, it seemed like a better performance to use a local cache rather than attempting to hit the api server so many times.

the reason for why we are hitting that api server so much is that we check for the scale from zero capability and limits many times within a single sync interval. if the community here is ok with our logs being full of client side throttling messages i can certainly switch this back to the other style, but the cache seemed to give a better user experience from the log perspective.

Reply:
>when running that way is that we get client side throttled several times per sync loop

Thanks for context Mike! Can you help me understand why de above happens? How's this different than any other resource we are consuming? why is a hand crafted cache justified only here? I think to mitigate overloading the api server we should be able to use existing kube core client cache mechanisms aka informers/indexers and if needed controller runtime abstractions . I'd be reluctant to have our own cache unless there's a strong reason for it.

Reply:
yeah, it's a great point Alberto, and i am not overly fond of creating a cache here but it seemed like the best way to handle this for the infrastructure templates only.

my rationale goes along these lines,
1. we need to discover the infrastructure template gvk and names from the machinesets and machinedeployments
2. adding the infrastructure templates to our dynamic informer caches does work, but given the frequency with which we check the scale from zero information, this produces a higher number of api calls than the calls to get the machinesets and machinedeployments
3. we expect a `capacity` field in the `status` for the infrastructure templates, in practice this capacity should not change for the life of the template. that said, it is possible that there might be a a period of time between when an infrastructure template is created and when the actuator adds the capacity information, so we need to be able to reconcile these just in case it does change.
4. given the statements above, it seems more reasonable to cache the information from the template rather than request new information on every sync interval, especially when multiple machinesets/machinedeployments could share the same template.

that's my thinking, if we all agree that the excess calls to the api server are ok, then i will change it back to my original design. but, i do feel that the local cache is appropriate here.

Reply:
>given the statements above, it seems more reasonable to cache the information from the template rather than request new information on every sync interval, especially when multiple machinesets/machinedeployments could share the same template.
that's my thinking, if we all agree that the excess calls to the api server are ok, then i will change it back to my original design. but, i do feel that the local cache is appropriate here.

I agree with the above, what I'm not seeing is why you need a handcrafted cache for that and to avoid excess calls to the api server rather than use built-in caches as everywhere else. Kube core go client and controller runtime give you caches that we already use here and widely in the ecosystem. Why can you use kubernetes informer/indexer and their higher abstractions with the resync config of your choice as usual?

Reply:
gonna reverse the order of your questions ;)

> Why can't you use kubernetes informer/indexer and their higher abstractions with the resync config of your choice as usual?

as far as i know nothing aside from my lack of knowledge around how to craft those caches. it wasn't clear to me how i would construct a cache through go-client where i could reduce the sync period. when i add the template types to the current client as we discover them, i see a lot of noise in the logs from the client blocking the calls due to the rate at which we are making calls. if those logs messages are acceptable, i will put the original client-go cache back in place.

> what I'm not seeing is why you need a handcrafted cache for that and to avoid excess calls to the api server rather than use built-in caches as everywhere else.

it was to reduce the log noise. using the built-in caches produces a lot of noise in the logs about `Waited for XXXms due to client-side throttling....`. is there a way to control how frequently the client-go client will connect with the API server?


and just to be clear, this is not a blocker for me. if we agree that the log noise is acceptable, i will revert to use the client-go cache.

Reply:
just to try and add a little more explanation,

the reason i felt the noise logs were an issue we should deal with is that we've had users question how the capi provider works with respect to seeing too many throttling messages in the logs. i don't want our users to think there is some problem that we could alleviate by reducing the frequency of our checks.

Reply:
If you are seeing `XXXms due to client-side throttling.....` every loop I question it's actually using the cache. This i happening now for e.g get scale subresource. See https://github.com/kubernetes/autoscaler/pull/5025

To move forward with this PR please let's move towards using native client/informers for cache.

Reply:
> If you are seeing `XXXms due to client-side throttling.....` every loop I question it's actually using the cache. This i happening now for e.g get scale subresource. See #5025

when i add my cache most of those messages go away for the infra templates

> To move forward with this PR please let's move towards using native client/informers for cache.

i'm ok with removing the additional cache, it just means our logs get more noisy. if that becomes an issue we can always put it back in place.

thanks for the followup =)

Reply:
>when i add my cache most of those messages go away for the infra templates

If you see it without your cache is because that client is hitting the api server instead of caching and that can be just codify to use the cache instead either in this PR or we can follow up later, let's just make sure we capture it in an issue.